{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 003 Minimal example of a rolling hash\n",
    "\n",
    "This notebook uses the low-level phylogenetic library [biomcmc-lib](https://github.com/quadram-institute-bioscience/biomcmc-lib) (commit [5975331](https://github.com/quadram-institute-bioscience/biomcmc-lib/commit/5975331ef88d1c4ec9aef9599fb6348905d289c7)).\n",
    "\n",
    "The rolling hash (or expoential hash) is a technique to slide through a sequence while updating its hash value using only information from the current hash and the next value. One of the most popular algorithms is the Rabin-Karp algorithm. Below I start with the imnplementation in [linclust/MMseqs2](http://dx.doi.org/10.1038/s41467-018-04964-5)\n",
    "\n",
    "In DNA context, each base is represented by a random number. In MMseqs they work with a reduced amino acid alphabet of 13 groups  \n",
    "\n",
    "## Prototype\n",
    "\n",
    "Here I describe a prototype (future capability of the library), not implemented yet. I am testing the rolling hash function but also shorter types `uint8_t` etc. The idea is that each of the testing functionalities will be incorporated into `biomcmc-lib`.\n",
    "\n",
    "Below is a working version that follows closely `minclust`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  0 >>  1 2 3  \t    23891  \t >>  1 2 3  \t    64755\n",
      "  1 >>  2 3 1  \t    35238  \t >>  2 3 4  \t    42449\n",
      "  2 >>  3 1 2  \t    11433  \t >>  3 4 5  \t    46207\n",
      "  3 >>  1 2 3  \t    23891  \t >>  4 5 2  \t    42985\n",
      "  4 >>  2 3 1  \t    35238  \t >>  5 2 3  \t    48371\n",
      "  5 >>  3 1 2  \t    11433  \t >>  2 3 4  \t    42449\n",
      "  6 >>  1 2 3  \t    23891  \t >>  3 4 5  \t    46207\n",
      "  7 >>  2 3 1  \t    35238  \t >>  4 5 2  \t    42985\n",
      "  8 >>  3 1 1  \t    38662  \t >>  5 2 3  \t    48371\n",
      "  9 >>  1 1 1  \t    17158  \t >>  2 3 4  \t    42449\n",
      " 10 >>  1 1 1  \t    17158  \t >>  3 4 5  \t    46207\n",
      " 11 >>  1 1 1  \t    17158  \t >>  4 5 1  \t     7238\n",
      " 12 >>  1 1 1  \t    17158  \t >>  5 1 1  \t      198\n",
      " 13 >>  1 1 1  \t    17158  \t >>  1 1 4  \t    48977\n",
      " 14 >>  1 1 2  \t    63657  \t >>  1 4 5  \t    62591\n",
      " 15 >>  1 2 1  \t    14054  \t >>  4 5 2  \t    42985\n",
      " 16 >>  2 1 2  \t    17577  \t >>  5 2 5  \t    43135\n",
      " 17 >>  1 2 3  \t    23891  \t >>  2 5 2  \t    42985\n",
      " 18 >>  2 3 1  \t    35238  \t >>  5 2 3  \t    48371\n",
      " 19 >>  3 1 2  \t    11433  \t >>  2 3 4  \t    42449\n",
      " 20 >>  1 2 1  \t    14054  \t >>  3 4 3  \t    41203\n",
      " 21 >>  2 1 2  \t    17577  \t >>  4 3 2  \t    57833\n",
      " 22 >>  1 2 1  \t    14054  \t >>  3 2 3  \t    48371\n",
      " 23 >>  2 1 2  \t    17577  \t >>  2 3 2  \t    57833\n",
      " 24 >>  1 2 1  \t    14054  \t >>  3 2 3  \t    48371\n"
     ]
    }
   ],
   "source": [
    "//%cflags:-lm\n",
    "//%cflags: -I/usr/users/QIB_fr005/deolivl/Academic/Quadram/009.supersptree/biomcmc-lib/lib\n",
    "//%cflags: -I/usr/users/QIB_fr005/deolivl/Academic/Quadram/009.supersptree/build/biomcmc-lib/lib\n",
    "//%cflags: /usr/users/QIB_fr005/deolivl/Academic/Quadram/009.supersptree/build/biomcmc-lib/lib/.libs/libbiomcmc.a\n",
    "#include <biomcmc.h>\n",
    "\n",
    "uint16_t RAND[21] = {0x4567, 0x23c6, 0x9869, 0x4873, 0xdc51, 0x5cff, 0x944a, 0x58ec,\n",
    "                           0x1f29, 0x7ccd, 0x58ba, 0xd7ab, 0x41f2, 0x1efb, 0xa9e3, 0xe146,\n",
    "                           0x007c, 0x62c2, 0x0854, 0x27f8, 0x231b};// 16 bit random numbers\n",
    "\n",
    "#define RoL(val, numbits) (val << numbits) ^ (val >> (32 - numbits))\n",
    "\n",
    "// Transform each letter x[i] to a fixed random number RAND[x[i]] to ensure instantaneous mixing into the 16 bits\n",
    "// Do XOR with RAND[x[i]] and 5-bit rotate left for each i from 1 to k\n",
    "uint32_t circ_hash (const uint8_t * x, uint8_t length, const uint8_t rol){\n",
    "    uint16_t h = 0x0;\n",
    "    h = h ^ RAND[x[0]];                  // XOR h and ki\n",
    "    for (uint8_t i = 1; i < length; ++i){\n",
    "        h = RoL(h, rol);\n",
    "        h ^= RAND[x[i]];                   // XOR h and ki\n",
    "    }\n",
    "    return h;\n",
    "}\n",
    "\n",
    "// Rolling hash for CRC hash variant: Computes hash value for next key x[0:length-1] from previous value\n",
    "// hash( x[-1:length-2] ) and x_first = x[-1]\n",
    "uint32_t circ_hash_next (const uint8_t * x, uint8_t length, uint8_t x_first, uint16_t h, const uint8_t rol){\n",
    "    // undo INITIAL_VALUE and first letter x[0] of old key\n",
    "    h ^= RoL(RAND[x_first], (5*(length-1)) % 16);// circularly permute all letters x[1:length-1] to 5 positions to left\n",
    "    // h ^= RoL(RAND[x_first], (5*(length-1)) & 15); // since x%y = x & (y-1)\n",
    "    h =  RoL(h, rol);// add new, last letter of new key x[1:length]\n",
    "    h ^= RAND[x[length-1]];\n",
    "    return h;\n",
    "}\n",
    "\n",
    "int main (){\n",
    "    uint8_t i, j;\n",
    "    uint8_t my_seq1[] = {1,2,3,1,2, 3,1,2,3,1, 1,1,1,1,1, 1,2,1,2,3, 1,2,1,2,1, 2,1,2}; // n=28\n",
    "    uint8_t my_seq2[] = {1,2,3,4,5, 2,3,4,5,2, 3,4,5,1,1, 4,5,2,5,2, 3,4,3,2,3, 2,3,2}; \n",
    "    uint16_t my_h1 = circ_hash((const uint8_t*) my_seq1, 3, 5);\n",
    "    uint16_t my_h2 = circ_hash((const uint8_t *) my_seq2, 3, 7);\n",
    "    for (j = 0; j < 25; j++) {\n",
    "        printf (\"%3d >>  \", j);\n",
    "        for (i=j;i<j+3;i++) printf (\"%d \", my_seq1[i]); \n",
    "        printf (\" \\t %8u  \\t >>  \",my_h1); // show current kmer and hash\n",
    "        for (i=j;i<j+3;i++) printf (\"%d \", my_seq2[i]); \n",
    "        printf (\" \\t %8u\\n\",my_h2); // show current kmer and hash\n",
    "        my_h1 = circ_hash_next ((const uint8_t *) my_seq1 + j + 1, 3, my_seq1[j], my_h1, 5);\n",
    "        my_h2 = circ_hash_next ((const uint8_t *) my_seq2 + j + 1, 3, my_seq2[j], my_h2, 7);\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The code above assumes a 16-bits hash, cast to 32 bits (due to the `RoL` macro). Below we use all 32 bits of hash, and also use the fact that for non-negative values, `x % y == x & (y-1)` which may be faster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  0 >>  1 2 3  10247507  \t >>  1 2 3 \t146668787\n",
      "  1 >>  2 3 1  40405414  \t >>  2 3 4 \t641639889\n",
      "  2 >>  3 1 2  19213481  \t >>  3 4 5 \t309507199\n",
      "  3 >>  1 2 3  10247507  \t >>  4 5 2 \t926590953\n",
      "  4 >>  2 3 1  40405414  \t >>  5 2 3 \t393460979\n",
      "  5 >>  3 1 2  19213481  \t >>  2 3 4 \t641639889\n",
      "  6 >>  1 2 3  10247507  \t >>  3 4 5 \t309507199\n",
      "  7 >>  2 3 1  40405414  \t >>  4 5 2 \t926590953\n",
      "  8 >>  3 1 1  19240710  \t >>  5 2 3 \t393460979\n",
      "  9 >>  1 1 1   9126662  \t >>  2 3 4 \t641639889\n",
      " 10 >>  1 1 1   9126662  \t >>  3 4 5 \t309507199\n",
      " 11 >>  1 1 1   9126662  \t >>  4 5 1 \t926555206\n",
      " 12 >>  1 1 1   9126662  \t >>  5 1 1 \t388890822\n",
      " 13 >>  1 1 1   9126662  \t >>  1 1 4 \t148946769\n",
      " 14 >>  1 1 2   9173161  \t >>  1 4 5 \t144700543\n",
      " 15 >>  1 2 1  10237670  \t >>  4 5 2 \t926590953\n",
      " 16 >>  2 1 2  40191145  \t >>  5 2 5 \t393455743\n",
      " 17 >>  1 2 3  10247507  \t >>  2 5 2 \t640985065\n",
      " 18 >>  2 3 1  40405414  \t >>  5 2 3 \t393460979\n",
      " 19 >>  3 1 2  19213481  \t >>  2 3 4 \t641639889\n",
      " 20 >>  1 2 1  10237670  \t >>  3 4 3 \t309502195\n",
      " 21 >>  2 1 2  40191145  \t >>  4 3 2 \t925950441\n",
      " 22 >>  1 2 1  10237670  \t >>  3 2 3 \t307281139\n",
      " 23 >>  2 1 2  40191145  \t >>  2 3 2 \t641655273\n",
      " 24 >>  1 2 1  10237670  \t >>  3 2 3 \t307281139\n"
     ]
    }
   ],
   "source": [
    "//%cflags:-lm\n",
    "//%cflags: -I/usr/users/QIB_fr005/deolivl/Academic/Quadram/009.supersptree/biomcmc-lib/lib\n",
    "//%cflags: -I/usr/users/QIB_fr005/deolivl/Academic/Quadram/009.supersptree/build/biomcmc-lib/lib\n",
    "//%cflags: /usr/users/QIB_fr005/deolivl/Academic/Quadram/009.supersptree/build/biomcmc-lib/lib/.libs/libbiomcmc.a\n",
    "#include <biomcmc.h>\n",
    "\n",
    "uint32_t RAND[21] = {0x4567, 0x23c6, 0x9869, 0x4873, 0xdc51, 0x5cff, 0x944a, 0x58ec,\n",
    "                           0x1f29, 0x7ccd, 0x58ba, 0xd7ab, 0x41f2, 0x1efb, 0xa9e3, 0xe146,\n",
    "                           0x007c, 0x62c2, 0x0854, 0x27f8, 0x231b};\n",
    "\n",
    "#define RoL(val, numbits) (val << numbits) ^ (val >> (32 - numbits))\n",
    "uint32_t circ_hash (const uint8_t * x, uint8_t length, const uint8_t rol){\n",
    "    uint32_t h = 0x0;\n",
    "    h = h ^ RAND[x[0]];                  // XOR h and ki\n",
    "    for (uint8_t i = 1; i < length; ++i){\n",
    "        h = RoL(h, rol);\n",
    "        h ^= RAND[x[i]];                 // XOR h and ki\n",
    "    }\n",
    "    return h;\n",
    "}\n",
    "\n",
    "uint32_t circ_hash_next (const uint8_t * x, uint8_t length, uint8_t x_first, uint32_t h, const uint8_t rol){\n",
    "    uint8_t remain = (rol * (length-1)) & 31U; // since x % y = x & (y-1)\n",
    "    // undo INITIAL_VALUE and first letter x[0] of old key\n",
    "    //h ^= RoL(RAND[x_first], (rol * (length-1)) % 32);// circularly permute all letters x[1:length-1] to 5 positions to left\n",
    "    h ^= RoL(RAND[x_first], remain);\n",
    "    h =  RoL(h, rol);// add new, last letter of new key x[1:length]\n",
    "    h ^= RAND[x[length-1]];\n",
    "    return h;\n",
    "}\n",
    "\n",
    "int main (){\n",
    "    uint8_t i, j;\n",
    "    uint8_t my_seq1[] = {1,2,3,1,2, 3,1,2,3,1, 1,1,1,1,1, 1,2,1,2,3, 1,2,1,2,1, 2,1,2}; // n=28\n",
    "    uint8_t my_seq2[] = {1,2,3,4,5, 2,3,4,5,2, 3,4,5,1,1, 4,5,2,5,2, 3,4,3,2,3, 2,3,2}; \n",
    "    uint32_t my_h1 = circ_hash((const uint8_t *) my_seq1, 3, 5);\n",
    "    uint32_t my_h2 = circ_hash((const uint8_t *) my_seq2, 3, 7);\n",
    "    for (j = 0; j < 25; j++) {\n",
    "        printf (\"%3d >>  \", j);\n",
    "        for (i=j;i<j+3;i++) printf (\"%d \", my_seq1[i]); \n",
    "        printf (\" %8u  \\t >>  \",my_h1); // show current kmer and hash\n",
    "        for (i=j;i<j+3;i++) printf (\"%d \", my_seq2[i]); \n",
    "        printf (\"\\t%8u\\n\",my_h2); // show current kmer and hash\n",
    "        my_h1 = circ_hash_next ((const uint8_t *) my_seq1 + j + 1, 3, my_seq1[j], my_h1, 5);\n",
    "        my_h2 = circ_hash_next ((const uint8_t *) my_seq2 + j + 1, 3, my_seq2[j], my_h2, 7);\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## transforming DNA bases into hash values\n",
    "\n",
    "Above, we have a single hash value `RAND[]` for each base, but we can create a stream of values.\n",
    "\n",
    "1. we define a translation table from DNA to $0...3$ for non-ambiguous data (or some other encoding). Notice that we must know its reverse complement.\n",
    "2. hash values inspired by `RAND[]` are then created. They are \"salted\" with large primes.\n",
    "\n",
    "The version currently implemented in `biomcmc-lib` is, for non-ambiguous DNA:\n",
    "```c\n",
    "uint8_t dna_in_2_bits[256][2] = {{0xff}};\n",
    "\n",
    "void\n",
    "initialize_dna_to_bit_tables (void) {\n",
    "    int i;\n",
    "    for (i = 0; i < 256; i++) dna_in_2_bits[i][0] = dna_in_2_bits[i][1] = 4; // calling function must check if < 4\n",
    "    dna_in_2_bits['A'][0] = 0;   dna_in_2_bits['A'][1] = 3;  /*  A  <-> T  */\n",
    "    dna_in_2_bits['C'][0] = 1;   dna_in_2_bits['C'][1] = 2;  /*  C  <-> G  */\n",
    "    dna_in_2_bits['G'][0] = 2;   dna_in_2_bits['G'][1] = 1;  /*  G  <-> C  */\n",
    "    dna_in_2_bits['T'][0] = 3;   dna_in_2_bits['T'][1] = 0;  /*  T  <-> A  */\n",
    "    dna_in_2_bits['U'][0] = 3;   dna_in_2_bits['U'][1] = 0;  /*  U  <-> A  */\n",
    "}\n",
    "```\n",
    "It just returns a small number, that is concatenated to form the k-mer. I.e. a sequence `ACGGT` becomes an integer with binary `00 01 10 10 01`. \n",
    "\n",
    "Below is the tentative code:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/tmpwprnos3t.c: In function ‘circular_hash_start’:\n",
      "/tmp/tmpwprnos3t.c:64:14: error: ‘RAND’ undeclared (first use in this function); did you mean ‘NAN’?\n",
      "         h ^= RAND[x[i]];                 // XOR h and ki\n",
      "              ^~~~\n",
      "              NAN\n",
      "/tmp/tmpwprnos3t.c:64:14: note: each undeclared identifier is reported only once for each function it appears in\n",
      "/tmp/tmpwprnos3t.c:64:19: error: ‘x’ undeclared (first use in this function)\n",
      "         h ^= RAND[x[i]];                 // XOR h and ki\n",
      "                   ^\n",
      "/tmp/tmpwprnos3t.c: In function ‘circular_hash_next’:\n",
      "/tmp/tmpwprnos3t.c:73:14: error: ‘RAND’ undeclared (first use in this function); did you mean ‘NAN’?\n",
      "     h ^= RoL(RAND[x_first], remain);\n",
      "              ^\n",
      "/tmp/tmpwprnos3t.c:53:29: note: in definition of macro ‘RoL’\n",
      " #define RoL(val, numbits) ((val) << (numbits)) ^ ((val) >> (32 - (numbits)))\n",
      "                             ^~~\n",
      "[C kernel] GCC exited with code 1, the executable will not be executed"
     ]
    }
   ],
   "source": [
    "//%cflags:-lm\n",
    "//%cflags: -I/usr/users/QIB_fr005/deolivl/Academic/Quadram/009.supersptree/biomcmc-lib/lib\n",
    "//%cflags: -I/usr/users/QIB_fr005/deolivl/Academic/Quadram/009.supersptree/build/biomcmc-lib/lib\n",
    "//%cflags: /usr/users/QIB_fr005/deolivl/Academic/Quadram/009.supersptree/build/biomcmc-lib/lib/.libs/libbiomcmc.a\n",
    "#include <biomcmc.h>\n",
    "const char dna[] = \n",
    "    \"AAACCACCTCCTCAATATATCCTAGATGTCATCATCTAAATACTCTAAATGAGATAACAG \\\n",
    "    GTTCTTAACATTCTCATAGGTTTCTCGTACAGCCACTCGTTTACCATCACTGGTGAAGAT \\\n",
    "    ACAGGTATTACCTTCGTCAACTGTAATGAATGTAATGTTATTGATATTAAACTCCACTGG \\\n",
    "    GTTATATAATGGTTGACTGTTTTGTTGATAACATATAGCCCCCTGTATTAGAGGGCTAGG \\\n",
    "    TTCATTGTTCTGTGTGGGGTTAGTCTCTTCTACATCAGCAGAGGTCTCTTCCTCGTATGC \\\n",
    "    TTCATACGGTACGTGTTCAGTAACAACAATCAAATTAGGAAACACTACAAGTTGTTCATC\";\n",
    "\n",
    "uint32_t rand_hash_list[] = {0x4567, 0x23c6, 0x9869, 0x4873, 0xdc51}; // size = 5\n",
    "uint32_t prime_salt_list[] = { // size = 64 \n",
    "  0x343EAF9F, 0x75BD32B,  0x5E1C5A87, 0x343EFDAF, 0x1FDCDBEF, 0x6389CB,   0x1FDCE507, 0x1FDCE3C7, \n",
    "  0x1FDCE15F, 0x75BD431,  0x34C8B0F,  0x5397FEF,  0x87413,    0x6389C9,   0x34C8B23,  0x343EFDB5, \n",
    "  0x75BD479,  0x5398013,  0x5397FCD,  0x87407,    0x75BD413,  0x343EAF4B, 0x343EB047, 0x6389CF, \n",
    "  0x5E1C5A2F, 0x343EFD31, 0x34C8B59,  0x343EFD57, 0x1FDCDC19, 0x5E1C5B0D, 0x34C8B5D,  0x343EB035, \n",
    "  0x75BD307,  0x343EAF7B, 0x343EAEF5, 0x75BD433,  0x1FDCDBD5, 0x1FDCE3B7, 0x343EFD19, 0x5398057, \n",
    "  0x34C8B51,  0x638971,   0x1FDCE16F, 0x63896F,   0x343EB04D, 0x873FD,    0x63896B,   0x1FDCDC9B, \n",
    "  0x5E1C5B1F, 0x343EFCEB, 0x5398007,  0x1FDCE51F, 0x63897B,   0x5E1C5B17, 0x5E1C5ABD, 0xFF85, \n",
    "  0x1FDCDC25, 0x873EB,    0x5E1C5AB1, 0x75BD30D,  0x1FDCDCC1, 0xFF8B,     0x5397FC1,  0x5398001};\n",
    "\n",
    "uint32_t**\n",
    "new_dna_salted_hash_encoding (uint8_t salt) {\n",
    "    uint8_t i, j;\n",
    "    uint32_t **shash = (uint32_t**) biomcmc_malloc (2 * sizeof (uint32_t*)); // opposite order as dna_in_2_bits[]\n",
    "    for (i = 0; i < 2;i++) shash[i] = (uint32_t*) biomcmc_malloc (256 * sizeof (uint32_t));\n",
    "    \n",
    "    for (i = 0; i < 256; i++) shash[0][i] = shash[1][i] = 4; // anything else treated as fifth state\n",
    "    shash[0]['A'] = shash[0]['a'] = 0; shash[1]['A'] = shash[1]['a'] = 3;  /*  A  <-> T  */\n",
    "    shash[0]['C'] = shash[0]['c'] = 1; shash[1]['C'] = shash[1]['c'] = 2;  /*  C  <-> G  */\n",
    "    shash[0]['G'] = shash[0]['g'] = 2; shash[1]['G'] = shash[1]['g'] = 1;  /*  G  <-> C  */\n",
    "    shash[0]['T'] = shash[0]['t'] = 3; shash[1]['T'] = shash[1]['t'] = 0;  /*  T  <-> A  */\n",
    "    shash[0]['U'] = shash[0]['u'] = 3; shash[1]['U'] = shash[1]['u'] = 0;  /*  U  <-> A  */\n",
    "    \n",
    "    salt &= 63; // modulus, we only have 64 salts\n",
    "    /** now we transform the indexes for their equiv. hash values; all have same salt */\n",
    "    for (j = 0; j < 2; ++j) for (i = 0; i < 256; ++i) shash[j][i] = rand_hash_list[shash[j][i]] + prime_salt_list[salt];\n",
    "    return shash;\n",
    "}\n",
    "\n",
    "void\n",
    "del_dna_salted_hash_encoding (uint32_t** shash) {\n",
    "    if (!shash) return;\n",
    "    if (shash[1]) free (shash[1]);\n",
    "    if (shash[0]) free (shash[0]);\n",
    "    free (shash);\n",
    "}\n",
    "\n",
    "\n",
    "#define RoL(val, numbits) ((val) << (numbits)) ^ ((val) >> (32 - (numbits)))\n",
    "// FIXME: code below incomplete \n",
    "uint32_t circular_hash_start (const char* dna, \n",
    "                              uint8_t kmer_length, \n",
    "                              const uint8_t rol, \n",
    "                              const uint32_t* shashcode, // just one of the two vectors\n",
    "                              const bool homopol_compress) {\n",
    "    uint32_t h = 0x0; \n",
    "    h = h ^ shashcode[(uint8_t)dna[0]];                  // XOR h and ki\n",
    "    for (uint8_t i = 1; i < kmer_length; ++i){\n",
    "        h = RoL(h, rol);\n",
    "        h ^= RAND[x[i]];                 // XOR h and ki\n",
    "    }\n",
    "    return h;\n",
    "}\n",
    "\n",
    "uint32_t circular_hash_next (const uint8_t * x, uint8_t length, uint8_t x_first, uint32_t h, const uint8_t rol){\n",
    "    uint8_t remain = (rol * (length-1)) & 31U; // since x % y = x & (y-1)\n",
    "    // undo INITIAL_VALUE and first letter x[0] of old key\n",
    "    //h ^= RoL(RAND[x_first], (rol * (length-1)) % 32);// circularly permute all letters x[1:length-1] to 5 positions to left\n",
    "    h ^= RoL(RAND[x_first], remain);\n",
    "    h =  RoL(h, rol);// add new, last letter of new key x[1:length]\n",
    "    h ^= RAND[x[length-1]];\n",
    "    return h;\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## conclusions\n",
    "\n",
    "`linclust` may have a \"bug\" since a five is hardcoded instead of `rol` in ` RoL(RAND[x_first], (5*(length-1)) % 16);`"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "C",
   "language": "c",
   "name": "c"
  },
  "language_info": {
   "file_extension": ".c",
   "mimetype": "text/plain",
   "name": "c"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
