{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 004 incorporating kalign3 as a library\n",
    "\n",
    "This notebook uses the low-level phylogenetic library [biomcmc-lib](https://github.com/quadram-institute-bioscience/biomcmc-lib) (commit [5975331](https://github.com/quadram-institute-bioscience/biomcmc-lib/commit/5975331ef88d1c4ec9aef9599fb6348905d289c7)).\n",
    "\n",
    "Recently I (re-)discovered the [kalign](https://github.com/TimoLassmann/kalign) software for multiple sequence alignment by Timo Lassmann, and was happily surprised with the code simplicity and a liberal license (GPL-3.0-or-later). \n",
    "\n",
    "Therefore I decided to incorporate its code into a library, that I can access using a `char_vector::` from the `biomcmc-lib` library. Currently I am working on a derived library/software called \"cumaru\" (still private), with a modified version of `kalign` algorithms.\n",
    "\n",
    "## Autotools\n",
    "\n",
    "First, a digresssion on the configuration aspects of `cumaru` (or another software importing these libraries).\n",
    "\n",
    "This is needed by `configure.ac` as well as the described files, in order for `kalign3` to use the AVX extensions. \n",
    "```bash\n",
    "# M4 macros for checking of CPU features from kalign3\n",
    "m4_include([m4/ax_gcc_x86_avx_xgetbv.m4])\n",
    "m4_include([m4/ax_gcc_x86_cpuid.m4])\n",
    "m4_include([m4/ax_check_compile_flag.m4])\n",
    "m4_include([m4/ax_ext.m4])\n",
    "m4_include([m4/ax_openmp.m4])\n",
    "AX_EXT\n",
    "```\n",
    "For importing `biomcmc-lib`, you have two options:\n",
    "1. downloading the (upstream) software through github, with option `--recursive` (default for final users of software). This will download `biomcmc-lib` into `${srcdir}/submodules/biomcmc-lib`.\n",
    "2. download `biomcmc-lib` independently into `${srcdir}/biomcmc-lib` or somewhere else and then link it to there. This is how I do it, since I use the same directory for several projects that rely on it. \n",
    "\n",
    "If you opted for (1) above (i.e. you just cloned the repository recursively), then the configuration option below will create a link as in option (2). \n",
    "\n",
    "```bash\n",
    "AC_CHECK_FILE([${srcdir}/biomcmc-lib],[],[ln -s submodules/biomcmc-lib ${srcdir}/biomcmc-lib])\n",
    "AC_CHECK_FILE([${srcdir}/biomcmc-lib/configure.ac],[], [AC_MSG_ERROR([\"biomcmc-lib submodule missing, please git clone --recursive or link by hand to location of source code\"])])\n",
    "dnl Call biomcmc-lib ./configure script recursively.\n",
    "AC_CONFIG_SUBDIRS([biomcmc-lib])\n",
    "AC_SUBST([BIOMCMCLIB], [biomcmc-lib])\n",
    "```\n",
    "\n",
    "Then the file `kalign/Makefile.am` has the following information, to create a static local library. This library will encapsulate `biomcmc-lib` and will be available to the main software as `libalign.la`:\n",
    "```bash\n",
    "AM_CPPFLAGS = $(GTKDEPS_CFLAGS)  -I$(srcdir)/../@BIOMCMCLIB@/lib  @OPENMP_CPPFLAGS@ @ZLIB_LIBS@\n",
    "AM_CFLAGS = @SIMD_FLAGS@ @AM_CFLAGS@  @OPENMP_CFLAGS@\n",
    "LDADD = $(GTKDEPS_LIBS) @ZLIB_LIBS@ ../biomcmc-lib/lib/libbiomcmc.la  $(AM_LDFLAGS)\n",
    "\n",
    "common_headers = kalign.h \\\n",
    "tldevel.h rng.h global.h \\\n",
    "alignment_parameters.h \\\n",
    "bisectingKmeans.h \\\n",
    "sequence_distance.h \\\n",
    "alignment.h bpm.h\n",
    "\n",
    "common_src = run_kalign.c \\\n",
    "tldevel.c rng.c \\\n",
    "alignment_parameters.c \\\n",
    "bisectingKmeans.c \\\n",
    "sequence_distance.c \\\n",
    "alignment.c bpm.c\n",
    "\n",
    "noinst_LTLIBRARIES = libkalign.la   ## noinst_LT: linked statically (not installed globally)\n",
    "libkalign_la_SOURCES = config.h $(common_headers) $(common_src) \n",
    "```\n",
    "\n",
    "And the `src/Makefile.am` (with the final software) can be something like:\n",
    "```bash\n",
    "AM_CPPFLAGS = $(GTKDEPS_CFLAGS) -I$(srcdir)/../kalign -I$(srcdir)/../@BIOMCMCLIB@/lib @OPENMP_CPPFLAGS@  @ZLIB_LIBS@ \n",
    "AM_CFLAGS = @AM_CFLAGS@ @OPENMP_CFLAGS@ @CHECK_CFLAGS@\n",
    "LDADD = $(GTKDEPS_LIBS) @CHECK_LIBS@  ../kalign/libkalign.la ../biomcmc-lib/lib/libbiomcmc.la @ZLIB_LIBS@  $(AM_LDFLAGS)\n",
    "\n",
    "bin_PROGRAMS = cumaru \n",
    "cumaru_SOURCES = main.c kseq.h\n",
    "cumaru_LDADD = $(LDADD)\n",
    "```\n",
    "Notice that we need to include the path to the local `biomcmc-lib` as well, since it is also statically linked.\n",
    "\n",
    "The code below is a minimal software that performs multiple sequence alignment from a fasta file (notice that the fasta reading comes from `kseq.h`). The code does not run, BTW. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/tmpscuivx3u.c:5:10: fatal error: kalign.h: No such file or directory\n",
      " #include <kalign.h>\n",
      "          ^~~~~~~~~~\n",
      "compilation terminated.\n",
      "[C kernel] GCC exited with code 1, the executable will not be executed"
     ]
    }
   ],
   "source": [
    "//%cflags:-lm\n",
    "//%cflags: -I/usr/users/QIB_fr005/deolivl/Academic/Quadram/009.supersptree/biomcmc-lib/lib\n",
    "//%cflags: -I/usr/users/QIB_fr005/deolivl/Academic/Quadram/009.supersptree/build/biomcmc-lib/lib\n",
    "//%cflags: /usr/users/QIB_fr005/deolivl/Academic/Quadram/009.supersptree/build/biomcmc-lib/lib/.libs/libbiomcmc.a\n",
    "#include <kalign.h>\n",
    "#include \"kseq.h\"\n",
    "KSEQ_INIT(gzFile, gzread)\n",
    "\n",
    "int\n",
    "main (int argc, char **argv)\n",
    "{\n",
    "  int i;\n",
    "  clock_t time0, time1;\n",
    "  char_vector seqname = new_char_vector (1);\n",
    "  char_vector dna = new_char_vector (1);\n",
    "  char_vector align = NULL;\n",
    "\n",
    "  time0 = clock ();\n",
    "  arg_parameters params = get_parameters_from_argv (argc, argv);\n",
    "\n",
    "  gzFile fp = gzopen((char*) params.fasta->filename[0], \"r\");\n",
    "  kseq_t *seq = kseq_init(fp);\n",
    "  while ((i = kseq_read(seq)) >= 0) {\n",
    "    char_vector_add_string (seqname, seq->name.s);\n",
    "    char_vector_add_string (dna, seq->seq.s);\n",
    "  }\n",
    "  kseq_destroy(seq);\n",
    "  gzclose(fp);\n",
    "  time1 = clock (); fprintf (stderr, \"read : %lf\\n\",  (double)(time1-time0)/(double)(CLOCKS_PER_SEC)); fflush(stderr); \n",
    "\n",
    "  align = kalign3_from_char_vector (dna);\n",
    "  for (i= 0; i < align->nstrings; i++) printf (\">%s\\n%s\\n\", seqname->string[i], align->string[i]);\n",
    "\n",
    "  del_char_vector (dna);\n",
    "  del_char_vector (align);\n",
    "  del_char_vector (seqname);\n",
    "  del_arg_parameters (params);\n",
    "  return EXIT_SUCCESS;\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "C",
   "language": "c",
   "name": "c"
  },
  "language_info": {
   "file_extension": ".c",
   "mimetype": "text/plain",
   "name": "c"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
